components/cosmetic_filters/renderer/cosmetic_filters_js_handler.cc

```diff
@@ -116,6 +116,51 @@ const char kHideSelectorsInjectScript[] =
           };
         })();)";
 
+const char kRemovalsInjectScript[] =
+    R"((function() {
+          const selectorsToRemove = %s;
+          const classesToRemoveBySelector = %s;
+          const attributesToRemoveBySelector = %s;
+          const onMutation = (mutationList) => {
+            if (mutationList.find(
+                mutation => (
+                  mutation.addedNodes.length !== 0 && mutation.addedNodes.values().find(node => node.nodeType == 1)
+                )
+              )
+            ) {
+              execute();
+            }
+          };
+          const execute = () => {
+              selectorsToRemove.forEach(
+                selector => document.querySelectorAll(selector).forEach(elem => elem.remove())
+              );
+              for (const selector in classesToRemoveBySelector) {
+                const classesToRemove = classesToRemoveBySelector[selector];
+                document.querySelectorAll(selector).forEach(elem => {
+                  const classesToActuallyRemove = classesToRemove.filter(elem.classList.contains.bind(elem.classList));
+                  if (classesToActuallyRemove.length) {
+                    elem.classList.remove(classesToActuallyRemove);
+                  }
+                });
+              }
+              for (const selector in attributesToRemoveBySelector) {
+                document.querySelectorAll(selector).forEach(elem =>
+                  attributesToRemoveBySelector[selector]
+                    .filter(elem.hasAttribute.bind(elem))
+                    .forEach(elem.removeAttribute.bind(elem))
+                );
+              }
+          };
+          const observer = new MutationObserver(onMutation);
```

Comment:
I strongly suggest to reuse the current observer [here](https://github.com/brave/brave-core/blob/master/components/cosmetic_filters/resources/data/content_cosmetic.ts#L190) because of perf reasons.

We also had perf issues with that one , so we had to tweak it adding a logic "too much mutations => switch to pooling (aka querySelectorAll by timer)"

Reply:
Hmm... OK I will look into this, though I'm a little unsure how to merge this in with the queue pumping. I'd set some global variables and the current observer script would call into some new removal-related function on mutation or on polling depending on the mode.

Reply:
AFAIK it doesn't directly connect with the queue pumping logic.
The mutation observer is created near the script loading:
`tryScheduleQueuePump()` (in the script end) =>`scheduleQueuePump()`

Reply:
The main issue with a new mutation observer that in corner cases (a lot ot mutations) it will produce a lot of work just to make a list with the updates on C++ side. Even the callback is about empty.

Anyway, I believe it's a good thing is to add some throttling to `execute()` callback (let's say limit it to few calls per seconds)

Reply:
I don't communicate back to C++...